csrc/fusion_segmenter.cpp

```diff
@@ -1784,77 +1786,23 @@ void eraseInputDistinctRootDomains(Fusion* fusion) {
       }
     }
 
-    TensorDomain* new_td = nullptr;
-    if (tv->domain()->hasAllocation()) {
-      // we need to reorder the logical domain into allocation domain
-      // consistently with the mapping from the old TensorView logical domain to
-      // its allocation domain
-      std::unordered_map<IterDomain*, IterDomain*> old_to_new;
-      for (const auto i : arange(logical.size())) {
-        old_to_new.emplace(logical[i], new_logical_domain[i]);
-      }
-
-      ReplayTransformations replay(tv->getAllocationDomain(), old_to_new);
-      // Without this,
-      // https://github.com/NVIDIA/Fuser/blob/e613929a6c21b3095c8817b01b8f177096a26e60/csrc/transform_iter.cpp#L299
-      // tries to look for root IDs in the map, which shouldn't exist because
-      // the whole purpose of this function is to remove the root domain.
-      replay.setErrorOnFailure(false);
-      // We don't need replay.setReplayRFactor(true). The new root is the same
-      // as the new logical so there aren't any expressions between them.
-
-      std::vector<IterDomain*> new_alloc;
-      new_alloc.reserve(tv->getAllocationDomain().size());
-      for (IterDomain* alloc_id : tv->getAllocationDomain()) {
-        IterDomain* new_alloc_id = replay.getReplay().at(alloc_id);
-        // ReplayTransformations replay transforms but not paralelization, so
-        // we have to manually parallelize the new allocation ID. In other
-        // places, parallelization is usually done through parallelizeAllLike.
-        new_alloc_id->parallelize(alloc_id->getParallelType());
-        new_alloc.push_back(new_alloc_id);
-      }
-
-      std::vector<IterDomain*> new_loop;
-      if (tv->getLoopDomain() == tv->getAllocationDomain()) {
-        new_loop = new_alloc;
-      } else {
-        new_loop = new_logical_domain;
-      }
-
-      new_td = IrBuilder::create<TensorDomain>(
-          /*root_domain=*/std::vector<IterDomain*>(),
-          new_logical_domain,
-          new_alloc,
-          new_loop,
-          tv->domain()->contiguity());
-    } else {
-      new_td = IrBuilder::create<TensorDomain>(
-          new_logical_domain, tv->domain()->contiguity());
-    }
-
-    // Remove reduction domains from new_td
-    if (new_td->hasReduction()) {
+    TensorDomain* new_td = IrBuilder::create<TensorDomain>(new_logical_domain);
+    TransformReplay::selfReplay(tv->domain(), new_td, true);
+    if (!tv->domain()->hasAllocation()) {
+      // The default contiguity for new_td is false. `selfReplay` does not
+      // replay contiguity when no allocation domain is present.
```

Comment:
Thanks for catching this! I believe this is a bug in selfReplay. I assumed contiguity and allocation domain appear hand in hand, but you proved me wrong. Feel free to stay as is -- I can fix this in a separate PR.

Reply:
I was in two minds about this as well.

At all other places, the tensordomain comes from the `ops` API, and hence has contiguity set based on the inputs. I can move it to within `selfReplay`.